opentelemetry-user-events-metrics/src/exporter/mod.rs

```diff
@@ -2,31 +2,40 @@ use crate::transform::transform_resource_metrics;
 use async_trait::async_trait;
 
 use opentelemetry_api::metrics::{MetricsError, Result};
-use opentelemetry_sdk::metrics::{
+use opentelemetry_sdk::{metrics::{
     data::{ResourceMetrics, Temporality},
     exporter::PushMetricsExporter,
     reader::{AggregationSelector, TemporalitySelector},
     Aggregation, InstrumentKind,
-};
+}};
 
 use crate::tracepoint;
 use prost::Message;
 use std::fmt::{Debug, Formatter};
+use std::pin::Pin;
+use eventheader::_internal as ehi;
 
-#[derive(Clone, Copy)]
-pub struct MetricsExporter {}
+pub struct MetricsExporter {
+    trace_point: Pin<Box<ehi::TracepointState>>
+}
 
 impl MetricsExporter {
     pub fn new() -> MetricsExporter {
+        let trace_point = Box::pin(ehi::TracepointState::new(0));
         // This is unsafe because if the code is used in a shared object,
         // the event MUST be unregistered before the shared object unloads.
         unsafe {
-            let result = tracepoint::register();
-            if result != 0 {
-                eprintln!("Tracepoint failed to register.");
+            let result = tracepoint::register(trace_point.as_ref());
+            // If result is -1, tracepoint might be in an erroneous state, attempt to retry
+            if result == -1 {
+                eprintln!("Tracepoint failed to register. Try restarting your application.");
+            } else if result != 0 {
```

Comment:
positive `result` is also considered failure?

Reply:
The panic catch returns -1 for panic. (Shouldn't happen.)

The actual API returns 0 for success and an errno value (positive) for error.

Reply:
Removed.